### ISSUE TITLE
How should I change the geometry of the objects in the code?

### ISSUE THREAD

**Issue by gaoyao-xt** [xcompact3d/Incompact3d#46]
I am a newbie of using this code.
I tried to rewrite "subroutine body(ux,uy,uz,ep1)" and rewrote it to the geometry I wanted.
 The original code is as follows:
 subroutine body(ux,uy,uz,ep1)
!******************************************************************
USE param 
USE decomp_2d
USE variables
USE IBM
implicit none
real(mytype), dimension(xstart(1):xend(1),xstart(2):xend(2),xstart(3):xend(3)) :: ux,uy,uz,ep1
integer :: i,j,k
real(mytype) :: xm,ym,r
ep1=0.
do k=xstart(3),xend(3)
do j=xstart(2),xend(2)
do i=xstart(1),xend(1)
   xm=(i-1)*dx 
   ym=yp(j)!(j-1)*dy
   r=sqrt((xm-cex)*(xm-cex)+(ym-cey)*(ym-cey)) 
   if (r-ra >= 0.) cycle
   ux(i,j,k)=0.
   uy(i,j,k)=0. 
   uz(i,j,k)=0.
   ep1(i,j,k)=1. 
enddo
enddo
enddo
return  
end subroutine body

The code after I changed it is like this:
subroutine body(ux,uy,uz,ep1)                                                  
!*******************************************************************
USE param 
USE decomp_2d
USE variables
USE IBM    
!****************************************************************************
implicit none
real(mytype), dimension(xstart(1):xend(1),xstart(2):xend(2),xstart(3):xend(3)) :: ux,uy,uz,ep1
integer :: i,j,k
integer ii,jj,nn,mm
real(mytype) :: xm,ym,r
real(8) xx11,xx22,yy11,yy22
real(8),ALLOCATABLE :: pointx(:),pointy(:)
!Read the airfoil data file "a.txt"
open(17,FILE='a.txt')
READ(17,*)mm,nn !nn=2
ALLOCATE(pointx(mm),pointy(mm))
DO ii=1,mm
READ(17,*)pointx(ii),pointy(ii)
ENDDO
close(17)  
   ep1=0.
do k=xstart(3),xend(3)
do j=xstart(2),xend(2)
do i=xstart(1),xend(1)
    xm=(i-1)*dx 
    ym=yp(j)!(j-1)*dy
!**********************************************************************
   do ii=1,mm   
   xx11=pointx(ii)
   yy11=pointy(ii)             
      if ( i < mm ) then
      xx22 = pointx(ii+1)
      yy22 = pointy(ii+1)
    else
      xx22 = pointx(1)
      yy22 = pointy(1)
    end if
      if ( ( yy11 < ym .eqv. ym <= yy22 ) .and. &
        ( xm - xx11 ) * abs ( yy22 - yy11 ) < ( xx22 - xx11 ) * ( ym - yy11 ) ) then
       ux(i,j,k)=0.
       uy(i,j,k)=0.
       uz(i,j,k)=0.
       ep1(i,j,k)=1.
       end if
    end do
return
end do
end do
end do        
return    
    end subroutine body

When my calculation is complete, the calculation results are not correct, so I want to ask if the geometry of the object is not modified here?

I hope to get your answer, thank you first.

**Comment by fschuch** [comment]
Hi,

What version of the code are you using? It looks like an old version of the `subroutine body`, and not the version that we have here in the repository.

Anyway, the `subroutine geomcomplex` is the right place to define the geometry for your application.

**Comment by gaoyao-xt** [comment]
dear Felipe,
First of all, thank you for your attention to my question.
The problem now is that when I read the new geometry, it doesn't show up in the calculation, which means there is no geometry in the calculation process.
I also modified this version downloaded from Incompact3d - Home (Incompact3d in a wake around a cylinder configuration with an explicit time advancement with Re=300.). When I modify it in my way, I can't display the geometry in the calculation.
The following two figures, Figure 1 is the calculation result of the "cylinder.tar" model, and Figure 2,Figure 3 are the result of my modification,and Figure 4 is my geometry. Whether to modify the geometry not only  to modify this place, but also to modify other places.

![YHO%4Q}8HCMX3%166J`_38L](https://user-images.githubusercontent.com/57210548/68096989-45036500-feef-11e9-99de-1efdec204526.png)
![BI5 G$HIWR97O@XW%W@@`H](https://user-images.githubusercontent.com/57210548/68097025-85fb7980-feef-11e9-80d0-c58cf2adbeba.png)
![R}JAWOHP R_C BK ~TY7G$L](https://user-images.githubusercontent.com/57210548/68097030-8f84e180-feef-11e9-8b0b-0457bb7b3a72.png)
![GD3T 7C6FXE}2K6P$W84AT1](https://user-images.githubusercontent.com/57210548/68097035-94e22c00-feef-11e9-9e0a-8e9efce9871e.png)

**Comment by slaizet** [comment]
Hi,
If you are using cylinder.tar, you need to follow the instructions from https://www.incompact3d.com/download.html:
--> First, compile and execute the file "genepsi3d.f90" ("gfortran -O3 genepsi3d.f90" and then "./a.out")
--> Then run incompact3d (with ilag=1 in the file "incompact3d.prm").

So the geometry is imposed in genepsi3d.f90, not in body.f90. 

However, as Felipe said, you should be using the latest version of the code:https://github.com/xcompact3d/Incompact3d

Best wishes,

Sylvain

**Comment by gaoyao-xt** [comment]
dear Sylvain
I just follow the instructions (https://www.incompact3d.com/download.html :), the code above is another version I tried.
I know that the geometry is in genepsi3d.f90. I found an airfoil data file, wrote the coordinates of the airfoil data file into a two-dimensional array, and then judged whether xm, ym are inside the airfoil. However, when the calculation is completed, the output image shows that my airfoil is not read, so I don't know how to modify the code.

*************************************************************************************

(1)The code after I changed it is like this:
subroutine cylinder(epsi,nx,ny,nz,dx,yp,dz,remp)
   implicit none
!
   real(4),dimension(nx,ny,nz) :: epsi
   real(4),dimension(      ny) :: yp
   integer                     :: nx,ny,nz
   real(4)                     :: dx,dz
   real(4)                     :: xc,yc,r
   real(4)                     :: remp
   integer                     :: i,j,k,icheck
   real(4)                     :: x,y,x0,y0,x1,y1,xm,ym,pi,xa,xlx,xy,x00
   real(4)                     :: alpha
   integer                        ii,jj,nn,mm
   real(4)                        xx11,xx22,yy11,yy22
   real(4),ALLOCATABLE         :: pointx(:),pointy(:)

pi=acos(-1.)

!Read airfoil data file
open(17,FILE='a.txt')
READ(17,*)mm,nn !nn=2
ALLOCATE(pointx(mm),pointy(mm))
DO ii=1,mm
READ(17,*)pointx(ii),pointy(ii)
ENDDO
close(17)
  
do k=1,nz
do j=1,ny
do i=1,nx
   xm=(i-1)*dx 
   ym=yp(j)
!**********************************************************************
!Determine if the point is inside the polygon
   do ii=1,mm   
   xx11=pointx(ii)
   yy11=pointy(ii)          
   
      if ( i < mm ) then
      xx22 = pointx(ii+1)
      yy22 = pointy(ii+1)
    else
      xx22 = pointx(1)
      yy22 = pointy(1)
    end if
      if ( ( yy11 < ym .eqv. ym <= yy22 ) .and. &
        ( xm - xx11 ) * abs ( yy22 - yy11 ) < ( xx22 - xx11 ) * ( ym - yy11 ) ) then
       epsi(i,j,k)=remp
    end if   
    end do
return
end do
end do
end do        
return    
    end subroutine cylinder

***********************************************************************************

(2)  The original code is as follows:
subroutine cylinder(epsi,nx,ny,nz,dx,yp,dz,remp)
   implicit none
!
   real(4),dimension(nx,ny,nz) :: epsi
   real(4),dimension(      ny) :: yp
   integer                     :: nx,ny,nz
   real(4)                     :: dx,dz
   real(4)                     :: xc,yc,r
   real(4)                     :: remp
   integer                     :: i,j,k,icheck
   real(4)                     :: x,y,x0,y0,x1,y1,xm,ym,pi,xa,xlx,xy,x00
   real(4)                     :: alpha

!
pi=acos(-1.)

do k=1,nz
do j=1,ny
do i=1,nx
   xm=(i-1)*dx 
   ym=yp(j)
   r=sqrt((xm-5.)*(xm-5.)+(ym-6.)*(ym-6.))
   if (r-0.5 >= 0.) cycle
   epsi(i,j,k)=remp
enddo
enddo
enddo
   return
end subroutine cylinder

********************************************************************************************

My English is relatively poor, so please don't mind if there were any improper expression in my word.

Best wishes,

gaoyao

**Comment by slaizet** [comment]
Dear gaoyao,
You are probably doing something wrong when reading/writing your data. As it is not a problem related to the code, we cannot assist you with it. My only recommendation would be to use equations (not tabulated data) for your airfloil; you will easily find equations for most of the well-known airfoils.
All the best,
Sylvain

**Comment by gaoyao-xt** [comment]
Dear Sylvain
1.What you mean is that I am doing the right thing, but my code may be wrong. So I just need to change my code to read the airfoil data correctly.
2.Today I changed the cylinder to a square with a side length of "1" and changed the governing equation to (4.5<xm<5.5 and 5.5<ym<6.5), which would give me a good result.

![Uploading 2019-11-05 13_56_59____________.png…]()


3.What you mean is to find the equation of the airfoils and then use the equation of the airfoils to find the point inside the airfoils.

Best wishes,
gaoyao

**Comment by gaoyao-xt** [comment]
I accidentally turned this off.

**Comment by slaizet** [comment]
Dear gaoyao,

Yes to your question 1 and to your question 2!

All the best,

Sylvain

**Comment by gaoyao-xt** [comment]
Dear Sylvain
I have some things lately, so I didn’t respond to you in time, sorry.

First of all, thank you for your help, I can change the geometry  correctly now.

I want to output the calculation results of lift and drag, but I can see that there is no such output in these models. What should I do?

Best wishes,
gaoyao

**Comment by slaizet** [comment]
Dear Gaoyao,

Lift and drag calculations are implemented in the latest github version of the code so you will have to use that version.

Best wishes,

Sylvain

**Comment by gaoyao-xt** [comment]
I have downloaded the latest version on github, it looks very different from the version I used before, I am familiar with it as soon as possible.
There is another problem, I want to calculate a dynamic geometry, whether this code can still be used normally (I see that the speed is set to zero when dealing with the geometry inside, as the post in the following figure), or you Any other suggestions?
![1%T@R80RCB@~ X1OV V~X 7](https://user-images.githubusercontent.com/57210548/69062378-cdd4e180-0a55-11ea-80b7-540e68c4a322.png)

**Comment by slaizet** [comment]
Hi Gaoyao,

In principle no issues to use a moving object (only issue would be if you have a passive scalar and no flux condition as explained by Felipe). You just need to update the position of the object at each time step and make sure to reconstruct the flow accordingly at each time step.

All the best,

Sylvain

**Comment by gaoyao-xt** [comment]
Dear Sylvain
I downloaded the latest version of Xcompact, but I had a problem when i am compiling, I think I have installed all the required dependencies (gfortran, openmpi, git...). Then I used the 'make' command but it went wrong.
I would like to ask, is it compiled and run the same as the previous version (make, mpirun -np Core-number ./xcompact3d, this is from the beginner's guide)
![QJ6Y{ % `@Z SNC0MBWSYRU](https://user-images.githubusercontent.com/57210548/69428848-29141600-0d6d-11ea-8250-419bd494218b.png)

Best wishes,
gaoyao

**Comment by ricardofrantz** [comment]
We have indeed to update the guide. Please try: mpirun -n 4 ./xcompact3d
examples/Turbulent-Boundary-Layer/input.i3d
Beware the number after -n must match p_row*p_col, which is set in the
first lines of input.i3d.

Em sex., 22 de nov. de 2019 às 14:15, gaoyao-xt <notifications@github.com>
escreveu:

**Comment by gaoyao-xt** [comment]
Dear Sylvain
Hello, the Linux version information I used is as follows:
![VP@9QNML%8YD)(X(4RIDVTR](https://user-images.githubusercontent.com/57210548/70425494-473b7f00-1aac-11ea-9f27-495426c91356.png)
But I haven't been able to install the software Git correctly. Is my system version too old? what should I do?
Best wishes,
gaoyao

**Comment by slaizet** [comment]
Hi Gaoyao,
We don't deal with Github queries but you can also download the zip file for xcompact3d.
Cheers,
Sylvain

**Comment by gaoyao-xt** [comment]
Dear Sylvain
I have download the zip file for xcompact3d,but when I compile the code(I use the MAKE command), it prompts as shown below:

![1](https://user-images.githubusercontent.com/57210548/70487896-c626cb00-1b31-11ea-9cf5-4dee65a55fb8.png)

It prompts:    git: Command not found
I change the p_row*p_col=2*4,so i use the command: 'mpirun -n 8 ./xcompact3d ./examples/Cylinder/input.i3d' 
But some errors occur at runtime, as shown below:
![3](https://user-images.githubusercontent.com/57210548/70488331-47cb2880-1b33-11ea-9228-680391686385.png)
Please help me check the problem
Best wishes,
Gaoyao

**Comment by gaoyao-xt** [comment]
Dear Sylvain
I have download the zip file for xcompact3d,but when I compile the code(I use the MAKE command), it prompts as shown below:

![1](https://user-images.githubusercontent.com/57210548/70490768-42bda780-1b3a-11ea-85ce-ab9a285b4792.png)

It prompts:    git: Command not found
I change the p_row*p_col=2*4,so i use the command: 'mpirun -n 8 ./xcompact3d ./examples/Cylinder/input.i3d' 
But some errors occur at runtime, as shown below:

![3](https://user-images.githubusercontent.com/57210548/70490800-55d07780-1b3a-11ea-887f-85be62394b6b.png)

Please help me check the problem
Best wishes,
Gaoyao

**Comment by ricardofrantz** [comment]
Well if you intent to run the code with -n 8 you should specify p_row and p_col aiming a product equal to 8. You could try 2 * 4 or 4 * 2 for instance. In my experience better performance is achieve with p_row ~ p_col and p_row < p_col. Keep in mind this is not a fixed rule. If you don’t know which combination to use and you don’t need optimal performance you can always start with 0 * 0 to activate ‘autotune’ and use any number of ranks available in your system. 
Ricardo

**Comment by gaoyao-xt** [comment]
Dear Ricardo
I make 'p_row =2 and p_col=4' but it doesn't work, it will have problems like the one I mentioned above (forrtl: error (78): process killed (SIGTERM)).
I don't know whata should i do

Best wishes,
Gaoyao

**Comment by slaizet** [comment]
Hi Gaoyao,

You can modify the Makefile to remove any git command. Regarding your issue when running the simulation, the code is working absolutely fine so if you follow the instruction from Ricardo it should work. Signal 11 suggests that you have a memory error so you should look in this direction. I am afraid we cannot help you more, on our side the code is working absolutely fine.

Best wishes,

Sylvain

**Comment by gaoyao-xt** [comment]
Dear Sylvain
You mean that the command shown in the figure below can be deleted, it does not affect the normal operation of the code.
![VHLNV53T`CRRY7}N23}O6CQ](https://user-images.githubusercontent.com/57210548/70539632-f062a200-1b9e-11ea-8245-a4d9ab19c2b7.png)

Best wishes,
Gaoyao

**Comment by slaizet** [comment]
just remove option -DVERSION=\"$(GIT_VERSION)\"

**Comment by gaoyao-xt** [comment]
![](https://user-images.githubusercontent.com/57210548/70539632-f062a200-1b9e-11ea-8245-a4d9ab19c2b7.png)
Does this need to be commented out?

**Comment by slaizet** [comment]
it does not matter!
just remove option -DVERSION="$(GIT_VERSION)"

**Comment by gaoyao-xt** [comment]
Dear Sylvain
Thank you very much for your guidance
Best wishes,
Gaoyao

**Comment by gaoyao-xt** [comment]
Dear Sylvain
The same compile and run, the example Channel-Flow works well
![WAH4LOTOHFT2{I@@ 2N{ JK](https://user-images.githubusercontent.com/57210548/70550710-23ae2c80-1bb1-11ea-9c98-41b77f141f02.png)

Best wishes,
Gaoyao

**Comment by gaoyao-xt** [comment]
Dear Sylvain
I have download the latest github version of the code and run it.
I want to output the calculation results of lift and drag but i can't find this option  :
![image](https://user-images.githubusercontent.com/57210548/70911591-2e534080-204d-11ea-9e0d-adf919c79700.png)
I use the example of Cylinder and I also don't see the lift and drag output options in the subroutine visu.f90
Can you tell me what should i do ?

Best wishes,
Gaoyao

**Comment by gaoyao-xt** [comment]
Dear Sylvain
I want to use a moving object ,I changed the subroutine '
![W)AB) Y5C{XT)PS_)U`F IR](https://user-images.githubusercontent.com/57210548/71062328-2dcdbd80-21a5-11ea-949e-3e7cebff7b4a.png)
' in the example 'BC-Cylinder.f90'.
I used ‘itime’ as a time variable to enable the model to move after 20,000 time steps, but it did n’t move.
But I changed 'genepsi3d.f90' or 'body.f90' in the old version of 'incompact3d' and it can move. 
Below is the code I changed:
![image](https://user-images.githubusercontent.com/57210548/71062490-9a48bc80-21a5-11ea-9611-a9eb46d6f267.png)

Best wishes,
Gaoyao
